beacon_node/network/src/subnet_service/mod.rs

```diff
@@ -17,12 +32,645 @@ pub enum SubnetServiceMessage {
+
+    /// A list of permanent subnets that this node is subscribed to.
+    // TODO: Shift this to a dynamic bitfield
+    permanent_attestation_subscriptions: HashSet<Subnet>,
+
+    /// A collection timeouts to track the existence of aggregate validator subscriptions at an
+    /// `ExactSubnet`.
+    aggregate_validators_on_subnet: Option<HashSetDelay<ExactSubnet>>,
+
+    /// The waker for the current thread.
+    waker: Option<std::task::Waker>,
+
+    /// The discovery mechanism of lighthouse is disabled.
+    discovery_disabled: bool,
+
+    /// We are always subscribed to all subnets.
+    subscribe_all_subnets: bool,
+
+    /// Whether this node is a block proposer-only node.
+    proposer_only: bool,
+
+    /// The logger for the attestation service.
+    log: slog::Logger,
+}
+
+impl<T: BeaconChainTypes> SubnetService<T> {
+    /* Public functions */
+
+    /// Establish the service based on the passed configuration.
+    pub fn new(
+        beacon_chain: Arc<BeaconChain<T>>,
+        node_id: NodeId,
+        config: &NetworkConfig,
+        log: &slog::Logger,
+    ) -> Self {
+        let log = log.new(o!("service" => "attestation_service"));
+
+        let slot_duration = beacon_chain.slot_clock.slot_duration();
+
+        if config.subscribe_all_subnets {
+            slog::info!(log, "Subscribing to all subnets");
+        }
+
+        // Build the list of known permanent subscriptions, so that we know not to subscribe or
+        // discover them.
+        let mut permanent_attestation_subscriptions = HashSet::default();
+        if config.subscribe_all_subnets {
+            // We are subscribed to all subnets, set all the bits to true.
+            for index in 0..<T::EthSpec as EthSpec>::SubnetBitfieldLength::to_u64() {
+                permanent_attestation_subscriptions
+                    .insert(Subnet::Attestation(SubnetId::from(index)));
+            }
+        } else {
+            // Not subscribed to all subnets, so just calculate the required subnets from the
+            for subnet_id in SubnetId::compute_attestation_subnets::<T::EthSpec>(
+                node_id.raw().into(),
+                &beacon_chain.spec,
+            ) {
+                permanent_attestation_subscriptions.insert(Subnet::Attestation(subnet_id));
+            }
+        }
+
+        // Set up the sync committee subscriptions
+        let spec = &beacon_chain.spec;
+        let epoch_duration_secs =
+            beacon_chain.slot_clock.slot_duration().as_secs() * T::EthSpec::slots_per_epoch();
+        let default_sync_committee_duration = Duration::from_secs(
+            epoch_duration_secs.saturating_mul(spec.epochs_per_sync_committee_period.as_u64()),
+        );
+
+        let track_validators = !config.import_all_attestations;
+        let aggregate_validators_on_subnet =
+            track_validators.then(|| HashSetDelay::new(slot_duration));
+
+        let mut events = VecDeque::with_capacity(10);
```

Comment:
do we need to allocate more capacity for events now the two services are merged? or are there less events because long live subscriptions are now permanant?

Reply:
Its kind of an arbitrary number. There's usually 3 or 4 at any given time (discoveries, subscriptions etc). But if the node is slow these could queue a bit. Figured 10 was reasonable. These messages are tiny in memory tho, so 10 is still very small.